examples/api/lib/material/context_menu/editable_text_toolbar_builder.1.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This example demonstrates showing a custom context menu only when some
+// narrowly defined text is selected.
+
+import 'package:flutter/material.dart';
+
+void main() => runApp(MyApp());
+
+const String emailAddress = 'me@example.com';
+const String text = 'Select the email address and open the menu: $emailAddress';
+
+class MyApp extends StatelessWidget {
+  MyApp({super.key});
+
+  final TextEditingController _controller = TextEditingController(
+    text: text,
+  );
+
+  void _showDialog (BuildContext context) {
+    Navigator.of(context).push(
+      DialogRoute<void>(
+        context: context,
+        builder: (BuildContext context) =>
+          const AlertDialog(title: Text('You clicked send email!')),
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      home: Scaffold(
+        appBar: AppBar(
+          title: const Text('Custom button for emails'),
+        ),
+        body: Center(
+          child: Column(
+            children: <Widget>[
+              Container(height: 20.0),
+              TextField(
+                controller: _controller,
+                contextMenuBuilder: (BuildContext context, EditableTextState editableTextState) {
+                  final List<ContextMenuButtonItem> buttonItems =
+                      editableTextState.contextMenuButtonItems;
```

Comment:
Also I think either `contextMenuButtonItems`
and `getAnchors` should both be in the EditableState, or both are coming from a static method of `AdaptiveTextSelectionToolbar`. It feels weird that one of them is a getter in editable, while the other one retrieved through AdaptiveTextSelectionToolbar. It will be good that all of information to customize the context menu is in a central place. 

In the future, the API to retrieve those information can also be a mixin so that any widget state use that mixin can have the context menu.

Reply:
About your first comment: Good idea, I'll keep an eye out for how users are using this after it's merged.  Your idea would make it easier to append buttons, which is probably the most common use case.  For fancier things like inserting buttons in other places, removing buttons, modifying buttons, they can still use the approach in this example.

About your second comment: I took your advice and moved the anchors to EditableTextState and SelectableRegionState, see commit https://github.com/flutter/flutter/pull/107193/commits/327d47bc8900271c00d4766bdb526ce9947af8f9.